sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala

```diff
@@ -400,7 +403,10 @@ case class Expm1(child: Expression) extends UnaryMathExpression(StrictMath.expm1
   """,
   since = "1.4.0")
 case class Floor(child: Expression) extends UnaryMathExpression(math.floor, "FLOOR") {
-  override def dataType: DataType = child.dataType match {
+
+  private lazy val childDataType = child.dataType
```

Comment:
This looks a pretty common code pattern, so could we move this into base classes, e.g., lazy val in `UnaryExpression`?